service/backend/pchain/indexer/parser.go

```diff
@@ -0,0 +1,318 @@
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByIndex(ctx, index-1)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) ParseBlockWithHash(ctx context.Context, hash string) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	hashID, err := ids.FromString(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByID(ctx, hashID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) parseBlockBytes(proposerBytes []byte) (*ParsedBlock, error) {
+	errs := wrappers.Errs{}
+
+	proposer, bytes, err := getProposerFromBytes(proposerBytes)
+	if err != nil {
+		return nil, fmt.Errorf("fetching proposer from block bytes errored with %w", err)
+	}
+
+	if p.genesisTimestamp.IsZero() {
+		return nil, errParserUninitialized
+	}
+
+	// Default block time to proposer timestamp if exists, otherwise to genesis block
+	blockTimestamp := new(time.Time)
+
+	if proposer.Timestamp > genesisTimestamp {
+		*blockTimestamp = time.Unix(proposer.Timestamp, 0)
+	} else {
+		*blockTimestamp = time.Unix(genesisTimestamp, 0)
+	}
+
+	var blk blocks.Block
+	ver, err := p.codec.Unmarshal(bytes, &blk)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshaling block bytes errored with %w", err)
+	}
+	blkID := ids.ID(hashing.ComputeHash256Array(bytes))
+
+	parsedBlock := ParsedBlock{
+		Height:    blk.Height(),
+		BlockID:   blkID,
+		BlockType: fmt.Sprintf("%T", blk),
+		Proposer:  proposer,
+	}
+
+	switch castBlk := blk.(type) {
+	case *blocks.ApricotProposalBlock:
+		errs.Add(common.InitializeTx(ver, p.codec, castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+	case *blocks.BlueberryProposalBlock:
+		errs.Add(common.InitializeTx(ver, p.codec, castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+	case *blocks.ApricotAtomicBlock:
+		errs.Add(common.InitializeTx(ver, p.codec, castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+	case *blocks.ApricotStandardBlock:
+		var blockTxs []*txs.Tx
+
+		for i, tx := range castBlk.Transactions {
+			errs.Add(common.InitializeTx(ver, p.codec, tx))
+			blockTxs = append(blockTxs, castBlk.Transactions[i])
+		}
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = blockTxs
+	case *blocks.BlueberryStandardBlock:
+		var blockTxs []*txs.Tx
+
+		for i, tx := range castBlk.Transactions {
+			errs.Add(common.InitializeTx(ver, p.codec, tx))
+			blockTxs = append(blockTxs, castBlk.Transactions[i])
+		}
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = blockTxs
+	case *blocks.ApricotAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	case *blocks.BlueberryAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
```

Comment:
insert: `parsedBlock.Timestamp = castBlk.Timestamp().Unix()`

Reply:
@bilgehansahin we are pulling blocks from avalanchego index API, which provide a timestamp for the block. Why don't we use that?

Reply:
AFAICT, the container timestamp refers to the time a given container is accepted by the indexer node, not the time that block was created or broadcasted. For example in my test node I have been using since June, all containers that was published before the node was started gets a timestamp of the initial sync of the node in June. 

We wanted the timestamps to be consistent across nodes / deployments and hence looked at the block time. Pre-Banff it was a hit-and-miss, so if a block had a missing timestamp we injected the so called genesis timestamp as timestamps are required by Rosetta spec and the validation requires a timestamp after year 2000. With Banff, it seems every block has a timestamp so we can get an accurate timestamp as far as I can tell.

Reply:
@bilgehansahin, you're right about container timestamp limitation; you are also right that for post Banff blocks, the block timestamp wil be the `chain Time` at which that block is generated, hence that time is guaranteed to be monotonic and agreed upon by any node in the network.
The reason I proposed the indexer timestamp alternative is that indexer guarantee a monotonic time pre-Banff and across Banff fork activation. On the contrary in the current solution, we cut and past times from different source with no guarantee of monotonicity.
I am not sure what are the business requirements here and I am favour of picking the solution that serve them the best.